File: weather.cpp
#include "pxt.h"
#include "WeatherService.h"

extern PeridoRESTClient *radioTxRx;
WeatherService *weatherService = NULL;
/**
 * share
 * use
 */

enum class WeatherLocationType
{
    city,
    postcode
};

//% color=243 weight=100 icon="\uf185" block="Weather"
namespace weather {

    int init() {
        if (radioTxRx == NULL)
            radioTxRx = new PeridoRESTClient(uBit.radio, uBit.messageBus, uBit.serial);

        if (weatherService == NULL)
            weatherService = new WeatherService(*radioTxRx);

        return 0;
    }

    //%
    String getTemperature(WeatherLocationType locationType, String location, String country)
    {
        init();
        ManagedString s = weatherService->getTemperature((int)locationType, MSTR(location), MSTR(country));
        return PSTR(s);
    }

    //%
    String getWindDirection(WeatherLocationType locationType, String location, String country)
    {
        init();
        WeatherServiceWind w = weatherService->getWind((int)locationType, MSTR(location), MSTR(country));
        return PSTR(w.direction);
    }

    //%
    String getWeatherForecast(WeatherLocationType locationType, String location, String country)
    {
        init();
        WeatherServiceForecastNow w = weatherService->getForecastNow((int)locationType, MSTR(location), MSTR(country));
        return PSTR(w.text);
    }

    //%
    String getWeatherForecastTomorrow(WeatherLocationType locationType, String location, String country)
    {
        init();
        WeatherServiceForecastTomorrow w = weatherService->getForecastTomorrow((int)locationType, MSTR(location), MSTR(country));
        return PSTR(w.text);
    }
}

File: weather.ts
//% color=243 weight=100 icon="\uf185" block="Weather"
namespace weather {

    /**
      * Get current temperature in city or by location postcode
      * @param weatherLocationType city or postcode; eg: WeatherLocationType.city
      * @param location city name or location postcode; eg: "London"
      */
    //% help=weather/get-temperature
    //% weight=96
    //% blockId=get_place_temp
    //% block="get|temperature|for %weatherLocationType | %location | %country" blockGap=8
    //% country.defl="GB"
    //% inlineInputMode=inline
    //% shim=weather::getTemperature
    export function getTemperature(weatherLocationType: WeatherLocationType, location: string, country: string) : string
    {
        return Math.floor(Math.random() * 25).toString();
    }

    /**
      * Get current wind direction in city or by location postcode
      * @param weatherLocationType city or postcode; eg: WeatherLocationType.city
      * @param location city name or location postcode; eg: "London"
      */
    //% help=weather/get-wind-direction
    //% weight=96
    //% blockId=get_wind_dir
    //% block="get|wind direction for %weatherLocationType | %location | %country" blockGap=8
    //% country.defl="GB"
    //% shim=weather::getWindDirection
    export function getWindDirection(weatherLocationType: WeatherLocationType, location: string, country: string) : string
    {
        let directions = ["N", "NW", "NE","S","SE","SW","E","W"]
        return directions[Math.floor(Math.random() * 7)]
    }

    /**
      * Get current weather forecast for city or by location postcode
      * @param weatherLocationType city or postcode; eg: WeatherLocationType.city
      * @param location city name or location postcode; eg: "London"
      */
    //% help=weather/get-weather-forecast
    //% weight=96
    //% blockId=get_forecast_now
    //% block="get|weather forecast for %weatherLocationType | %location | %country" blockGap=8
    //% country.defl="GB"
    //% shim=weather::getWeatherForecast
    export function getWeatherForecast(weatherLocationType: WeatherLocationType, location: string, country: string) : string
    {
        let weather = ["overcast", "raining", "sunny", "snowing", "thunder"]
        return weather[Math.floor(Math.random() * 4)]
    }

    /**
      * Get tomorrows weather forecast for city or by location postcode
      * @param weatherLocationType city or postcode; eg: WeatherLocationType.city
      * @param location city name or location postcode; eg: "London"
      */
    //% help=weather/get-weather-forecast-tomorrow
    //% weight=96
    //% blockId=get_forecast_tom block="get|tomorrow's weather forecast for %weatherLocationType | %location | %country" blockGap=8
    //% country.defl="GB"
    //% shim=weather::getWeatherForecastTomorrow
    export function getWeatherForecastTomorrow(weatherLocationType: WeatherLocationType, location: string, country: string)
    {
        let weather = ["overcast", "raining", "sunny", "snowing", "thunder"]
        return weather[Math.floor(Math.random() * 4)]
    }
}